internal/controller/podschedulingpolicy_controller.go

```diff
@@ -0,0 +1,211 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package controllers contains a set of controllers for everest
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	everestv1alpha1 "github.com/percona/everest-operator/api/v1alpha1"
+	"github.com/percona/everest-operator/internal/controller/common"
+)
+
+// PodSchedulingPolicyReconciler reconciles a PodSchedulingPolicy object.
+type PodSchedulingPolicyReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=everest.percona.com,resources=podschedulingpolicies,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=everest.percona.com,resources=podschedulingpolicies/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// Modify the Reconcile function to compare the state specified by
+// the PodSchedulingPolicy object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *PodSchedulingPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (rr ctrl.Result, rerr error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling", "request", req)
+	defer func() {
+		logger.Info("Reconciled", "request", req)
+	}()
+
+	psp := &everestv1alpha1.PodSchedulingPolicy{}
+	if err := r.Get(ctx, req.NamespacedName, psp); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	dbList, err := common.DatabaseClustersThatReferenceObject(ctx, r.Client, podSchedulingPolicyNameField, "", psp.GetName())
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("Failed to fetch DB clusters that use pod scheduling policy='%s'", psp.GetName()))
+		return ctrl.Result{}, errors.Join(err, fmt.Errorf("failed to fetch DB clusters that use pod scheduling policy='%s': %w", psp.GetName(), err))
+	}
+
+	// Update the status and finalizers of the PodSchedulingPolicy object after the reconciliation.
+	defer func() {
+		// Nothing to process on delete events
+		if !psp.GetDeletionTimestamp().IsZero() {
+			return
+		}
```

Comment:
Why do we do this? It means that if there is a deletion timestamp, we will not update the status

Reply:
Is it really required to update the state of resource that is being deleted anyway? I just think this is not needed at all.

Reply:
> Is it really required to update the state of resource that is being deleted anyway?
Some CRDs have a `deleting` state, so if there is additional clean-up logic needed, you would need to reconcile a deleting object. But in this case, we don't have such a requirement, so I think it is fine. I just wanted to ensure that this was intentional